src/swupd/fullfiles.go

```diff
@@ -0,0 +1,229 @@
+		// NOTE: to make life simpler for the client, always use .tar extension even if the
+		// file could be compressed.
+		output := filepath.Join(outputDir, name+".tar")
+
+		var err error
+		switch f.Type {
+		case typeDirectory:
+			err = createDirectoryFullfile(input, name, output)
+		case typeLink:
+			err = createLinkFullfile(input, name, output)
+		case typeFile:
+			err = createRegularFullfile(input, name, output)
+		default:
+			return fmt.Errorf("cannot create fullfiles: file %s is of unsupported type %q", f.Name, f.Type)
+		}
+
+		if err != nil {
+			return err
+		}
+
+		done[f.Hash] = true
+	}
+	return nil
+}
+
+func createDirectoryFullfile(input, name, output string) error {
+	fi, err := os.Lstat(input)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+	if !fi.IsDir() {
+		return fmt.Errorf("couldn't create fullfile from %s: manifest expected a directory but it is not", input)
+	}
+
+	hdr, err := getHeaderFromFileInfo(fi)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+	hdr.Name = name
+	hdr.Typeflag = tar.TypeDir
+
+	err = createTarGzipHeaderOnly(output, hdr)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+
+	return nil
+}
+
+func createLinkFullfile(input, name, output string) error {
+	fi, err := os.Lstat(input)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+	if fi.Mode()&os.ModeSymlink == 0 {
+		return fmt.Errorf("couldn't create fullfile from %s: manifest expected a link but it is not", input)
+	}
+	link, err := os.Readlink(input)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+
+	hdr, err := getHeaderFromFileInfo(fi)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+	hdr.Name = name
+	hdr.Typeflag = tar.TypeSymlink
+	hdr.Linkname = link
+
+	err = createTarGzipHeaderOnly(output, hdr)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+
+	return nil
+}
+
+func createRegularFullfile(input, name, output string) error {
+	fi, err := os.Lstat(input)
+	if err != nil {
+		return fmt.Errorf("couldn't create fullfile from %s: %s", input, err)
+	}
+	if !fi.Mode().IsRegular() {
+		return fmt.Errorf("couldn't create fullfile from %s: manifest expected a regular file but it is not", input)
+	}
+
+	best := ""
+	var bestSize int64 = math.MaxInt64
+	for i, c := range fullfileCompressors {
+		candidate := fmt.Sprintf("%s.%d", output, i)
+		size, err := c.Func(input, name, candidate, fi)
+		if err != nil {
+			log.Printf("WARNING: couldn't compress %s using compressor %q: %s", input, c.Name, err)
+			os.RemoveAll(candidate)
+			continue
+		}
+		if size < bestSize {
+			if best != "" {
+				os.RemoveAll(best)
+			}
+			best = candidate
+			bestSize = size
+		} else {
+			os.RemoveAll(candidate)
+		}
+	}
+
```

Comment:
Before sending this PR I've experimented with a tar-only "compressor", and the tar-header overhead took away any advantage of not compressing the small files. My understanding is that tar is always necessary to keep the metadata unless we make some other change elsewhere.

Reply:
I expect that any compressor will manage to compress the tar header,  I agree that we need (at the moment) to keep the tar wrapper around the file for the meta data.

Notation. I am going to use file.tar.gz for the gzipped version of the tarfile containing file, although I know it will actually be called file.tar eventually.

My point is that we should be comparing file.tar with  file.tar.gz, file.tar.xz, and file.tar.bz, rather than just comparing the latter 3 among themselves  . If you have an empty list of compressors then file.tar is the winner as there are no better compressed versions (as there are no compressed versions at all). At the moment if you have an empty list of compressors you get an error for everything, whilst I assert that if you don't have any compressors the best you can do is use the file.tar as it is.

Does this make sense?

Reply:
After some thought, I agree. Specially because we are getting rid of an "error scenario", since there is always a valid file to use.

I wasn't even considering "tar only" because of the tests I did (mostly the files from os-core, os-core-update, kernel-native) and because the current swupd-server doesn't do that. Since swupd-client does "tar xf" I think we will be fine having the tar only files.

I rewrote that part of the code a little bit, to reuse the uncompressed tar file (taking advantage of it being a special thing, and not a "no compression" compressor). This will not serve for cases we are calling tar externally, but helps a little bit for the Go-only compressors.

PS: I'm being all "hey there can be externals" because we haven't validated the Go implementation tar yet, but I'm confident we will make things work without using the external tar tool or some C tar library implementation.